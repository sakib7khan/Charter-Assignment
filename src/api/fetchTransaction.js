export const fetchTransactions = async () => {
  const res = await fetch('http://localhost:3005/transactions');
  if (!res.ok) throw new Error('API fetch failed');
  return await res.json();
};
