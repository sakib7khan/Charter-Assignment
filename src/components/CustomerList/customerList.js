import React, { useEffect, useState } from 'react';
import { fetchTransactions } from '../../api/fetchTransaction';
import { Container, Title, Button, Spinner, ErrorMessage } from '../../styles/stylesComponents';

const CustomerList = ({ onSelectCustomer, selectedCustomer }) => {
  const [customers, setCustomers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchTransactions()
      .then((data) => {
        const uniqueCustomers = Array.from(
          new Map(data.map((txn) => [txn.customerId, { id: txn.customerId, name: txn.customerName }])).values()
        );
        setCustomers(uniqueCustomers);
        setLoading(false);
      })
      .catch((err) => {
        setError('Error loading customers.');
        setLoading(false);
      });
  }, []);

  if (loading) return <Spinner>Loading customers...</Spinner>;
  if (error) return <ErrorMessage>{error}</ErrorMessage>;

  return (
    <Container data-testid="customer-list">
      <Title>Select Customer</Title>
      {customers.map((customer) => (
        <Button
          key={customer.id}
          onClick={() => onSelectCustomer(customer.id)}
          active={selectedCustomer === customer.id}
          data-testid= "customer-button"
        >
          {customer.name}
        </Button>
      ))}
    </Container>
  );
};

export default CustomerList;
