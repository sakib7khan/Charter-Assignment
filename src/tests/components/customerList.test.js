import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import CustomerList from '../../components/CustomerList/customerList';
import { fetchTransactions } from '../../api/fetchTransaction';

jest.mock('../../api/fetchTransaction');

const mockTransactions = [
  { customerId: '1', customerName: 'Alice', amount: 120, date: '2024-01-01' },
  { customerId: '2', customerName: 'Bob', amount: 90, date: '2024-01-02' },
  { customerId: '1', customerName: 'Alice', amount: 50, date: '2024-01-03' },
];

describe('CustomerList Component', () => {
  it('should show loading initially', () => {
    fetchTransactions.mockReturnValue(new Promise(() => {})); // pending promise
    render(<CustomerList onSelectCustomer={jest.fn()} selectedCustomer={null} />);
    expect(screen.getByText('Loading customers...')).toBeInTheDocument();
  });

  it('should show error if fetch fails', async () => {
    fetchTransactions.mockRejectedValueOnce(new Error('Failed'));
    render(<CustomerList onSelectCustomer={jest.fn()} selectedCustomer={null} />);
    const error = await screen.findByText('Error loading customers.');
    expect(error).toBeInTheDocument();
  });

  it("matches the snapshot", () => {
    fetchTransactions.mockResolvedValueOnce([]);
    const { asFragment } = render(<CustomerList onSelectCustomer={jest.fn()} selectedCustomer={null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render unique customers from transactions', async () => {
    fetchTransactions.mockResolvedValueOnce(mockTransactions);
    render(<CustomerList onSelectCustomer={jest.fn()} selectedCustomer={null} />);
    await waitFor(() => {
      const buttons = screen.getAllByTestId('customer-button');
      expect(buttons.length).toBe(2); // Alice and Bob (unique)
      expect(screen.getByText('Alice')).toBeInTheDocument();
      expect(screen.getByText('Bob')).toBeInTheDocument();
    });
  });

  it('should trigger onSelectCustomer when button clicked', async () => {
    const mockSelect = jest.fn();
    fetchTransactions.mockResolvedValueOnce(mockTransactions);
    render(<CustomerList onSelectCustomer={mockSelect} selectedCustomer={null} />);
    const buttons = await screen.findAllByTestId('customer-button');
    fireEvent.click(buttons[1]); // Click Bob
    expect(mockSelect).toHaveBeenCalledWith('2');
  });
});
