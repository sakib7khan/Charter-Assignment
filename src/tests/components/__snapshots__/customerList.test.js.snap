// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerList Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iQQCXo bezgzY"
  >
    Loading customers...
  </div>
</DocumentFragment>
`;
